torch/csrc/jit/passes/concat_opt.cpp

```diff
@@ -0,0 +1,480 @@
+    }
+    for (auto n : nodes_to_remove_) {
+      removeCatNodeFromGraph(n);
+    }
+  }
+
+  void moveBefore(Node* node, Node* before) {
+    // In order to move a node before another node, we need to move
+    // all the nodes it depends on as well.
+    for (auto inp : node->inputs()) {
+      moveBefore(inp->node(), before);
+    }
+    node->moveBefore(before);
+  }
+
+  void reuseBuffersInCopies() {
+    for (auto copy : copies_added_) {
+      auto src = copy->input(1);
+      auto dst = copy->input(0);
+      if (src->node()->kind() != aten::empty) {
+        continue;
+      }
+
+      // Move the destination node before the source.
+      GRAPH_UPDATE("Moving\n", *dst->node(), "before\n", *src->node());
+      moveBefore(dst->node(), src->node());
+
+      GRAPH_UPDATE("Replacing\n", *src->node(), "with\n", *dst->node());
+      src->replaceAllUsesWith(dst);
+
+      GRAPH_UPDATE("Deleting\n", *src->node());
+      src->node()->destroy();
+
+      GRAPH_UPDATE("Deleting\n", *copy);
+      copy->destroy();
+    }
+  }
+
+  void removeRedundantSlices() {
+    for (auto slice : slices_added_) {
+      auto inp = slice->input(0)->node();
+      if (inp->kind() != aten::slice) {
+        continue;
+      }
+
+      bool redundant = true;
+      for (size_t i = 1; i < slice->inputs().size(); ++i) {
+        auto slice_inp = slice->input(i);
+        auto inp_inp = inp->input(i);
+        if (slice_inp == inp_inp) {
+          continue;
+        }
+        if (slice_inp->node()->hasAttribute(attr::value) &&
+            inp_inp->node()->hasAttribute(attr::value)) {
+          if (slice_inp->node()->i(attr::value) !=
+              inp_inp->node()->i(attr::value)) {
+            redundant = false;
+            break;
+          }
+        } else {
+          redundant = false;
+          break;
+        }
+      }
+      if (redundant) {
+        // Remove this slice and replace its uses with the input.
+        GRAPH_UPDATE("Replacing uses of\n", *slice, "with\n", *inp);
+        slice->replaceAllUsesWith(inp);
+
+        GRAPH_UPDATE("Deleting\n", *slice);
+        slice->destroy();
+      }
+    }
+  }
+
+  std::shared_ptr<Graph> graph_;
+  std::unordered_set<Node*> nodes_to_remove_;
+  std::unordered_map<Value*, Value*> replace_uses_with_;
+  std::vector<Node*> copies_added_;
+  std::vector<Node*> slices_added_;
+};
+
+} // namespace
+
+void EliminateConcatCse(const std::shared_ptr<Graph>& graph) {
+  ConcatCseEliminator(graph).run();
+  GRAPH_DUMP("After ConcatCSE", graph);
+}
+
+void ExpandConcatAndEliminateRedundancy(const std::shared_ptr<Graph>& graph) {
+  ConcatExpander(graph).run();
+  GRAPH_DUMP("After expanding Concat and eliminating redundancy", graph);
+}
+
+void OptimizeConcat(const std::shared_ptr<Graph>& graph) {
+  GRAPH_DUMP("Before ConcatOpt", graph);
+  EliminateConcatCse(graph);
+  ExpandConcatAndEliminateRedundancy(graph);
+  ConstantPooling(graph);
+  EliminateDeadCode(graph);
+  GRAPH_DUMP("After ConcatOpt", graph);
```

Comment:
Do we need to update aliasdb after node deletion and generating slice and copy_ like ops @eellison ?

Reply:
AliasDb is instantiated for each pass, not maintained globally so no need to do any work here